bridges/snowbridge/primitives/core/src/lib.rs

```diff
@@ -163,4 +165,48 @@ impl DescribeLocation for DescribeHere {
 
 /// Creates an AgentId from a Location. An AgentId is a unique mapping to a Agent contract on
 /// Ethereum which acts as the sovereign account for the Location.
-pub type AgentIdOf = HashedDescription<H256, (DescribeHere, DescribeFamily<DescribeAllTerminal>)>;
+pub type AgentIdOf =
+	HashedDescription<AgentId, (DescribeHere, DescribeFamily<DescribeAllTerminal>)>;
+
+#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct AssetMetadata {
+	pub name: BoundedVec<u8, ConstU32<32>>,
+	pub symbol: BoundedVec<u8, ConstU32<32>>,
+	pub decimals: u8,
+}
+
+pub type TokenId = H256;
+
+pub type TokenIdOf = HashedDescription<TokenId, DescribeGlobalPrefix<DescribeInner>>;
+
+pub struct DescribeGlobalPrefix<DescribeInterior>(sp_std::marker::PhantomData<DescribeInterior>);
+impl<Suffix: DescribeLocation> DescribeLocation for DescribeGlobalPrefix<Suffix> {
+	fn describe_location(l: &Location) -> Option<Vec<u8>> {
+		match (l.parent_count(), l.first_interior()) {
+			(1, Some(GlobalConsensus(network))) => {
+				let tail = l.clone().split_first_interior().0;
+				let interior = Suffix::describe_location(&tail.into())?;
+				Some((b"pna", network, interior).encode())
+			},
+			_ => None,
+		}
+	}
+}
+
+pub struct DescribeInner;
+impl DescribeLocation for DescribeInner {
+	fn describe_location(l: &Location) -> Option<Vec<u8>> {
+		match l.unpack().1 {
+			[] => Some(Vec::<u8>::new().encode()),
+			[Parachain(id)] => Some((*id).encode()),
+			[Parachain(id), PalletInstance(instance)] => Some((*id, *instance).encode()),
+			[Parachain(id), PalletInstance(instance), GeneralIndex(index)] =>
+				Some((*id, *instance, *index).encode()),
+			[Parachain(id), PalletInstance(instance), GeneralKey { data, .. }] =>
+				Some((*id, *instance, *data).encode()),
+			[Parachain(id), GeneralIndex(index)] => Some((*id, *index).encode()),
+			[Parachain(id), GeneralKey { data, .. }] => Some((*id, *data).encode()),
+			_ => None,
```

Comment:
This catch-all could break things if the location does not match  the previous patterns.

Suppose two Moonbeam tokens don't match, then the catch-all will end up allocating them the same token ID, causing massive breakage.

Perhaps we should pre-verify the token locations to ensure their locations are supported?

Reply:
Is there any reason we can't use `DescribeFamily<DescribeAllTerminal>` to cover all the variations (like we do for computing `AgentId`)

Reply:
`DescribeFamily` just does not work for the reachored location in format of ```(1,[GlobalConsensus(Network)...])```

Reply:
I think we need a comprehensive location hasher that handles all kinds of junctions. This will have less sharp edges for users, and is more future proof.

Surely this can be implemented using a simple iterative/recursive algorithm similar to DescribeFamily?

Reply:
Review of this PR is approaching head of my todo priority queue, will have more insight then, BUT from a cursory look it seems we should be enhancing `DescribeFamily<DescribeAllTerminal>` to cover the missing desired variations.

Reply:
`DescribeFamily` and related traits only works for Locations relative to the current chain. We currently use Locations relative to the Ethereum to generate the ID. This is probably the correct thing to do if we want to double bridge assets such as in the case of allowing Kusama assets to be bridged to Polkadot and then to Ethereum.

Reply:
Another reason why I think we should implement our own trait is that It becomes a gate to which tokens we will allowed to be registered. We can start off restrictive and open it to more cases.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/5546/commits/4090be1ea34de666029ccf44528231b799418223

Reply:
I am not sure if these new traits belong in `DescribeAllTerminal` which describes Locations which are accounts/sovereigns. The new traits we are adding describe Locations which are assets. Since we use `DescribeAllTerminal` to create channels for example, including assets here would allow assets to have channels for instance.